---
title: "Figure_2_Analysis"
author: "Sheldon Ding"
date: "2025-03-20"
output: html_document
---

```{r load libraries}
#packages (some packages might require dependent packages)
library(Matrix)
library(Seurat)
library(Signac)
library(patchwork)
library(tidyverse)
library(dplyr)
library(readr)
library(magrittr)
library(harmony)
library(clustree)
library(GenomicRanges)
library(msigdbr)
library(clusterProfiler)
library(AnnotationHub)
library(BSgenome.Hsapiens.UCSC.hg38)
library(IRanges)
library(Biostrings)
library(openxlsx)
library(biomaRt)
library(ComplexHeatmap)
```

```{r set seed}
set.seed(6)
```

```{r load object}
sobj <- readRDS("output/objects/atlas.RDS")
```

###Figure 2
```{r Figure 2A}
##subset Radial Glials and Astrocytes for further analysis
RGL_Astrocyte <- subset(sobj, idents = c("Radial Glia", "Astrocyte"))

##run harmony to remove batch effect again on RNA assay
RGL_Astrocyte_harmony <- RGL_Astrocyte#create a separate object in case mess up something
DefaultAssay(RGL_Astrocyte_harmony) <- "RNA"
RGL_Astrocyte_harmony <- RGL_Astrocyte_harmony %>% NormalizeData() %>%FindVariableFeatures(selection.method = "vst", nfeatures=3000) %>%   ScaleData() %>%SCTransform(vars.to.regress = NULL)
RGL_Astrocyte_harmony <- RunPCA(RGL_Astrocyte_harmony, assay = "SCT", npcs = 50)
RGL_Astrocyte_harmony <- RunHarmony(RGL_Astrocyte_harmony,group.by.vars = c("sample_id"),reduction = "pca", assay.use = "SCT", reduction.save = "rna_harmony")

#normalize and SCT transform again after harmony
DefaultAssay(RGL_Astrocyte_harmony) <- "RNA"
RGL_Astrocyte_harmony <- RGL_Astrocyte_harmony %>%
  NormalizeData() %>%
  FindVariableFeatures(selection.method = "vst", nfeatures=3000) %>%
  ScaleData() %>%
  SCTransform(vars.to.regress = NULL)

#recluster RGL_Astrocyte subset
RGL_Astrocyte_harmony <- RGL_Astrocyte_harmony  %>%
  RunUMAP(reduction = "rna_harmony", assay = "SCT", dims = 1:50, reduction.name = 'umap.rna', reduction.key = 'rnaUMAP_') %>%
  FindNeighbors(reduction = "rna_harmony") %>%
  FindClusters(resolution = c(0.2, 0.4, 0.6, 0.8, 1.0, 1.2), verbose = TRUE)

##run harmony to remove batch effect again on ATAC assay
DefaultAssay(RGL_Astrocyte_harmony) <- "ATAC" 
RGL_Astrocyte_harmony <- RunTFIDF(RGL_Astrocyte_harmony) 
RGL_Astrocyte_harmony <- FindTopFeatures(RGL_Astrocyte_harmony, min.cutoff = 'q0') 
RGL_Astrocyte_harmony <- RunSVD(RGL_Astrocyte_harmony) 
RGL_Astrocyte_harmony <- RunUMAP(RGL_Astrocyte_harmony, dims = 2:50, reduction = 'lsi')
RGL_Astrocyte_harmony <- RunHarmony(object = RGL_Astrocyte_harmony, group.by.vars = c('sample_id'),reduction = 'lsi', assay.use = 'ATAC', project.dim = FALSE,  reduction.save = "atac_harmony")
RGL_Astrocyte_harmony <- RunUMAP(RGL_Astrocyte_harmony, dims = 2:50, reduction = 'atac_harmony', reduction.name = 'umap.atac', reduction.key = 'atacUMAP_')

#to assgin color to each age group
devcolor <- c("Fetal" =  "#0D0887", "Infant" = "#7E03A8", "Child" ="#CC4678","Adolescent" = "#F89441","Adult" = "#F0F921")

##to plot RGL_Astrocyte_harmony age on RNA DimPlot
DimPlot(RGL_Astrocyte_harmony, group.by = "dev.group",reduction = "umap.rna", label = F, cols = devcolor)

##to determine SNN resolution and number of clusters
clustree(RGL_Astrocyte_harmony) #determined the SNN resolution @0.6

##to calculate differentially expressed genes (DEG) between different clusters
DefaultAssay(RGL_Astrocyte_harmony) <- "RNA"
RGL_Astrocyte_harmony_cluster_DEG <- FindAllMarkers(RGL_Astrocyte_harmony, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)

##to calculate differentially accessible peaks (DAP) between different clusters
DefaultAssay(RGL_Astrocyte_harmony) <- "ATAC"
RGL_Astrocyte_harmony_cluster_DAP <- FindAllMarkers(RGL_Astrocyte_harmony, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)

##based on cluster DAP, DEG and canonical marker (see Figure 2B analysis in the next "r chunk"), we determined cell type of each cluster
#set active identity to SCT_snn_res.0.6
RGL_Astrocyte_harmony <- SetIdent(RGL_Astrocyte_harmony, value = RGL_Astrocyte_harmony@meta.data$SCT_snn_res.0.6)
#set cluster identity to cell type
{
  cluster_0 <- "SEMA3E Astrocyte"
  cluster_1 <- "ROBO2 Radial Glia"
  cluster_2 <- "ROBO2 Radial Glia"
  cluster_3 <- "ROBO2 Radial Glia"
  cluster_4 <- "ROBO2 Astrocyte"
  cluster_5 <- "ROBO2 Astrocyte"
  cluster_6 <- "SLIT2 Radial Glia"
  cluster_7 <- "ROBO2 Radial Glia"
  cluster_8 <- "OL-like Astrocyte"
  cluster_9 <- "SLIT2 Radial Glia"
  cluster_10 <- "SEMA3E Astrocyte"
  cluster_11 <- "SEMA3E Astrocyte"
  cluster_12 <- "ROBO2 Astrocyte"
  cluster_13 <- "Undetermined Astrocyte"
}

new.cluster.ids <- c(cluster_0, cluster_1, cluster_2, cluster_3, cluster_4, cluster_5, cluster_6, cluster_7, cluster_8, cluster_9, cluster_10, cluster_11, cluster_12, cluster_13)
names(new.cluster.ids) <- levels(RGL_Astrocyte_harmony)
RGL_Astrocyte_harmony <- RenameIdents(RGL_Astrocyte_harmony, new.cluster.ids)
RGL_Astrocyte_harmony$Subcluster <- RGL_Astrocyte_harmony@active.ident
RGL_Astrocyte_harmony@meta.data$Subcluster <- factor(x = RGL_Astrocyte_harmony@meta.data$Subcluster, levels = c("ROBO2 Radial Glia", "SLIT2 Radial Glia", "ROBO2 Astrocyte","SEMA3E Astrocyte","OL-like Astrocyte","Undetermined Astrocyte"))

#assign colors to different subclusters
RGL_Astrocyte_subcluster_color = c("ROBO2 Radial Glia" = "#e2e700","SLIT2 Radial Glia" = "#FFDAB9", "ROBO2 Astrocyte" = "#FFA500","SEMA3E Astrocyte" = "#FF4F00","OL-like Astrocyte" = "#C2A162","Undetermined Astrocyte" = "#BDC3C7")

##to plot subcluster on RNA DimPlot
DimPlot(RGL_Astrocyte_harmony, label = F, reduction = "umap.rna", cols = RGL_Astrocyte_subcluster_color)
```

```{r Figure 2B}
##we selected Top5 DEG between different clusters and canonical markers
DefaultAssay(RGL_Astrocyte_harmony) <- "RNA"
RGL_Astrocyte_DotPlot_Genes <- read.xlsx("output/tables/RGL_Astrocyte_DotPlot_Genes.xlsx")

##to plot Top5 DEG between different clusters and canonical markers on DotPlot
DotPlot(RGL_Astrocyte_harmony, features = unique(RGL_Astrocyte_DotPlot_Genes$gene), group.by = "Subcluster") + theme(axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=0.5)) + scale_y_discrete(limits=rev(c("ROBO2 Radial Glia", "SLIT2 Radial Glia", "ROBO2 Astrocyte","SEMA3E Astrocyte","OL-like Astrocyte","Undetermined Astrocyte")))
```

```{r Figure 2C}
##to plot expression of "H3-3A" and "H3-3B" across subcluster
VlnPlot(RGL_Astrocyte_harmony, features = c("H3-3A","H3-3B"), group.by = "Subcluster", cols = RGL_Astrocyte_subcluster_color)

##to plot expression of "H3-3A" and "H3-3B" across age
VlnPlot(RGL_Astrocyte_harmony, features = c("H3-3A","H3-3B"), group.by = "dev.group", cols = devcolor)
```

```{r Figure 2D}
##Load the genesets
ensembl = useMart("ensembl", dataset = "hsapiens_gene_ensembl")

# For C5 gene ontology genesets
C5genesets <- msigdbr(species = "Homo sapiens", category = "C5")
C5togene <- C5genesets[,c("gs_exact_source","gene_symbol")]
C5toname <- C5genesets[,c("gs_exact_source","gs_name")]

# For GO Biological Process
GOgenesets <- subset(C5genesets, subset=C5genesets[,"gs_subcat"]=="GO:BP")
GOtogene <- GOgenesets[,c("gs_exact_source","gene_symbol")]
GOtoname <- GOgenesets[,c("gs_exact_source","gs_name")]

#extract DEG of each subcluster from RGL_Astrocyte_harmony_cluster_DEG and make a list:
list <- c("ROBO2_Radial_Glia", "SLIT2_Radial_Glia", "ROBO2_Astrocyte","SEMA3E_Astrocyte","OL_like_Astrocyte","Undetermined_Astrocyte")

#get the GO Biological Process
GO <- lapply(list, function (x)
{
  print(x)
  y <- enricher(gene = rownames(eval(parse(text = x))),
                pvalueCutoff = 0.05,
                pAdjustMethod = "BH",
                minGSSize = 15,
                maxGSSize = 1000,
                qvalueCutoff = 0.2,
                TERM2GENE = GOtogene,
                TERM2NAME = GOtoname)
})

#GO Top5 Heatmap
names(GO) <- list
Desc_list <- lapply(list, function(x){
df <- GO[[x]]
df <- df[df$p.adjust < 0.05,]
df <- df[order(df$p.adjust, decreasing = FALSE),]
desc <- df$Description
desc <- desc[1:5]
return(desc)
}
)
Desc <- unlist(Desc_list) %>% unique()
Desc.df <- as.data.frame(Desc)
Desc.df <- na.omit(Desc.df)
rownames(Desc.df) <- Desc.df$Desc
Desc <- na.omit(Desc)
cols <-lapply(list, function(subcluster){
  Desc.df[subcluster] <- 0
  Subcluster.Desc.df <- data.frame(GO[[subcluster]])
  Subcluster.Desc.df$qvalue.scaled <- -log(Subcluster.Desc.df$qvalue)
  sig <-data.frame(Subcluster.Desc.df[Subcluster.Desc.df$p.adjust < 0.05,]) 
  sig <- sig[order(sig$p.adjust, decreasing = F),]
  Desc.df[sig$Description,subcluster]<- Subcluster.Desc.df[Subcluster.Desc.df$Description %in% sig$Description, "qvalue.scaled"]
  Desc.df <- Desc.df[unique(Desc),] 
  rownames(Desc.df) <- Desc.df$Desc
  Desc.df$Desc <- NULL
  return(Desc.df)
  })
df2 <- data.frame(cols)
matrix_GO <- as.matrix(df2)
colnames(matrix_GO) <- colnames(df2)
rownames(matrix_GO) <- unique(Desc)
matrix_GO = t(apply(df2, 1, scale))

#annotation
AnnoDF <- data.frame(c("ROBO2 Radial Glia", "SLIT2 Radial Glia", "ROBO2 Astrocyte","SEMA3E Astrocyte","Oligo-like Astrocyte","Undetermined Astrocyte"))
colnames(AnnoDF) <- c("Cluster")
AnnoDF$Cluster <- factor(AnnoDF$Cluster, levels = c("ROBO2 Radial Glia", "SLIT2 Radial Glia", "ROBO2 Astrocyte","SEMA3E Astrocyte","OL-like Astrocyte","Undetermined Astrocyte"))
colAnno <- HeatmapAnnotation(
  df=AnnoDF,
  show_annotation_name = c(TRUE,TRUE),
  annotation_name_gp = gpar(fontsize = 12),
  annotation_name_side = "left",
  col = list(Cluster = c("ROBO2 Radial Glia" = "#e2e700","SLIT2 Radial Glia" = "#FFDAB9", "ROBO2 Astrocyte" = "#FFA500","SEMA3E Astrocyte" = "#FF4F00","OL-like Astrocyte" = "#C2A162","Undetermined Astrocyte" = "#BDC3C7")))

#draw the heatmap of GO biological processes difference between different subclusters
plot_GO <- Heatmap(matrix_GO, cluster_columns=F,cluster_rows =FALSE, row_order = Desc, row_names_gp = gpar(fontsize = 5),top_annotation = colAnno, show_heatmap_legend = F)
draw(plot_GO, heatmap_legend_side="left", annotation_legend_side="left")
```

```{r Figure 2E}
##calculate the DAP in each subcluster
DefaultAssay(RGL_Astrocyte_harmony) <- "ATAC"
RGL_Astrocyte_harmony <- SetIdent(RGL_Astrocyte_harmony, value = RGL_Astrocyte_harmony@meta.data$Subcluster)
RGL_Astrocyte_harmony_subcluster_DAP <- FindAllMarkers(RGL_Astrocyte_harmony, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)

#extract the DAP in each subcluster and make a list
RGL_Astrocyte_Subcluster_DAP_list <- list(ROBO2_Radial_Glia_DAP,SLIT2_Radial_Glia_DAP,ROBO2_Astrocyte_DAP,SEMA3E_Astrocyte_DAP,Oligo_like_Astrocyte_DAP)

#run the rGreat analysis
RGL_Astrocyte_Subcluster_DAP_list <- lapply(RGL_Astrocyte_Subcluster_DAP_list, function(x) { x <- x %>%
  separate(gene, into = c("seqnames", "start", "end"), sep = "-")
x$start <- as.numeric(x$start)
x$end <- as.numeric(x$end)
x$strand <- c("*")
return(x)})
g.RGL_Astrocyte_Subcluster_DAP_list <- lapply(RGL_Astrocyte_Subcluster_DAP_list, function(x) {x <- GenomicRanges::GRanges(
  seqnames = x$seqnames,
  ranges = IRanges(start = x$start,
                   end = x$end
  ),
  strand = x$strand,
  id = x$cluster,
  rank = x$avg_log2FC,
  adjust = x$p_val_adj)
return(x)})

#calculate the unique DAP in each subcluster
laps_Full <- findOverlapsOfPeaks(g.RGL_Astrocyte_Subcluster_DAP[[1]], g.RGL_Astrocyte_Subcluster_DAP[[2]], g.RGL_Astrocyte_Subcluster_DAP[[3]], g.RGL_Astrocyte_Subcluster_DAP[[4]], g.RGL_Astrocyte_Subcluster_DAP[[5]])

#visualize the unique DAP in each subcluster using VennDiagram
makeVennDiagram(laps_Full, NameOfPeaks = c("ROBO2 Radial Glia", "SLIT2 Radial Glia", "ROBO2 Astrocyte","SEMA3E Astrocyte","OL-like Astrocyte","Undetermined Astrocyte"), fill=c("ROBO2 Radial Glia" = "#e2e700","SLIT2 Radial Glia" = "#FFDAB9", "ROBO2 Astrocyte" = "#FFA500","SEMA3E Astrocyte" = "#FF4F00","OL-like Astrocyte" = "#C2A162","Undetermined Astrocyte" = "#BDC3C7"))

#create a list of subclusters for barplot
subcluster <- c("ROBO2 Radial Glia", "SLIT2 Radial Glia", "ROBO2 Astrocyte","SEMA3E Astrocyte","OL-like Astrocyte","Undetermined Astrocyte")

#get the number of unique peaks across subcluster from RGL_Astrocyte_harmony_cluster_DAP
RGL_Astrocyte_subcluster_unique_DAP <- c(6266, 915, 29, 181,281,0)

#create a dataframe
RGL_Astrocyte_subcluster_unique_DAP_df <- data.frame(subcluster, RGL_Astrocyte_subcluster_unique_DAP,stringsAsFactors = FALSE)
RGL_Astrocyte_subcluster_unique_DAP_df$subcluster <- factor(x = RGL_Astrocyte_subcluster_unique_DAP_df$subcluster, levels = c("ROBO2 Radial Glia", "SLIT2 Radial Glia", "ROBO2 Astrocyte","SEMA3E Astrocyte","OL-like Astrocyte","Undetermined Astrocyte"))

##to plot unique DAP across subclusters on BarPlot
ggplot() +
  geom_bar(data = RGL_Astrocyte_subcluster_unique_DAP_df, aes(x = subcluster, y = RGL_Astrocyte_subcluster_unique_DAP, fill = subcluster), stat = "identity") + scale_fill_manual(values = RGL_Astrocyte_subcluster_color) +
    scale_y_continuous(expand = c(0, 0)) +
    theme_classic() + 
  theme( axis.title.y = element_blank(), axis.title.x = element_blank())
```

```{r Figure 2F}
##calculate the DAP in each age group
DefaultAssay(RGL_Astrocyte_harmony) <- "ATAC"
RGL_Astrocyte_harmony <- SetIdent(RGL_Astrocyte_harmony, value = RGL_Astrocyte_harmony@meta.data$dev.group)
RGL_Astrocyte_harmony_age_DAP <- FindAllMarkers(RGL_Astrocyte_harmony, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)

#extract the DAP in each subcluster and make a list
RGL_Astrocyte_age_DAP_list <- list(Fetal_DAP, Infant_DAP, Child_DAP, Adolescent_DAP, Adult_DAP)

#run the rGreat analysis
RGL_Astrocyte_age_DAP_list <- lapply(RGL_Astrocyte_age_DAP_list, function(x) { x <- x %>%
  separate(gene, into = c("seqnames", "start", "end"), sep = "-")
x$start <- as.numeric(x$start)
x$end <- as.numeric(x$end)
x$strand <- c("*")
return(x)})
g.RGL_Astrocyte_age_DAP_list <- lapply(RGL_Astrocyte_age_DAP_list, function(x) {x <- GenomicRanges::GRanges(
  seqnames = x$seqnames,
  ranges = IRanges(start = x$start,
                   end = x$end
  ),
  strand = x$strand,
  id = x$cluster,
  rank = x$avg_log2FC,
  adjust = x$p_val_adj)
return(x)})

#calculate the unique DAP in each subcluster
laps_Full <- findOverlapsOfPeaks(g.RGL_Astrocyte_age_DAP_list[[1]], g.RGL_Astrocyte_age_DAP_list[[2]], g.RGL_Astrocyte_age_DAP_list[[3]], g.RGL_Astrocyte_age_DAP_list[[4]], g.RGL_Astrocyte_age_DAP_list[[5]])

#visualize the unique DAP in each subcluster using VennDiagram
makeVennDiagram(laps_Full, NameOfPeaks = c("Fetal", "Infant", "Child", "Adolescent", "Adult"), fill=c("Fetal" =  "#0D0887", "Infant" = "#7E03A8", "Child" ="#CC4678","Adolescent" = "#F89441","Adult" = "#F0F921"))

#create a list of age groups for barplot
age <- c("Fetal", "Infant", "Child", "Adolescent", "Adult")

#get the number of unique peaks across age from laps_Full
RGL_Astrocyte_age_unique_DAP <- c(10222, 697, 0, 429, 23)

#create a dataframe
RGL_Astrocyte_age_unique_DAP_df <- data.frame(age, RGL_Astrocyte_age_unique_DAP_df,stringsAsFactors = FALSE)
RGL_Astrocyte_age_unique_DAP_df$age <- factor(x = RGL_Astrocyte_age_unique_DAP_df$age, levels = c("Fetal", "Infant", "Child", "Adolescent", "Adult"))

##to plot unique DAP across age groups on BarPlot
ggplot() +
  geom_bar(data = RGL_Astrocyte_age_unique_DAP_df, aes(x = age, y = RGL_Astrocyte_age_unique_DAP, fill = age), stat = "identity") + scale_fill_manual(values = devcolor) +
    scale_y_continuous(expand = c(0, 0)) +
    theme_classic() + 
  theme( axis.title.y = element_blank(), axis.title.x = element_blank())
```

```{r}
saveRDS(RGL_Astrocyte_harmony, "/objects/RGL_Astrocyte_harmony.rds")
```




