---
title: "Figure_5_Analysis"
author: "Sheldon Ding"
date: "2025-03-21"
output: html_document
---

```{r load libraries}
#packages (some packages might require dependent packages)
library(Matrix)
library(Seurat)
library(Signac)
library(patchwork)
library(tidyverse)
library(dplyr)
library(readr)
library(magrittr)
library(harmony)
library(clustree)
library(GenomicRanges)
library(msigdbr)
library(clusterProfiler)
library(BSgenome.Hsapiens.UCSC.hg38)
library(IRanges)
library(Biostrings)
library(biomaRt)
library(DESeq2)
library(EnhancedVolcano)
```

```{r set seed}
set.seed(6)
```

```{r load object}
sobj <- readRDS("output/objects/atlas.RDS")
```

```{r set color for age}
devcolor <- c("Fetal" =  "#0D0887", "Infant" = "#7E03A8", "Child" ="#CC4678","Adolescent" = "#F89441","Adult" = "#F0F921")
```


###Figure 5
```{r}
##subset Immune cells for further analysis
Immune_subset <- subset(sobj, idents = c("Immune"))

##run harmony to remove batch effect again on RNA assay
Immune_harmony <- Immune_subset#create a separate object in case mess up something
DefaultAssay(Immune_harmony) <- "RNA"
Immune_harmony <- Immune_harmony %>% NormalizeData() %>%FindVariableFeatures(selection.method = "vst", nfeatures=3000) %>%   ScaleData() %>%SCTransform(vars.to.regress = NULL)
Immune_harmony <- RunPCA(Immune_harmony, assay = "SCT", npcs = 50)
Immune_harmony <- RunHarmony(Immune_harmony,group.by.vars = c("sample_id"),reduction = "pca", assay.use = "SCT", reduction.save = "rna_harmony")

#normalize and SCT transform again after harmony
DefaultAssay(Immune_harmony) <- "RNA"
Immune_harmony <- Immune_harmony %>%
  NormalizeData() %>%
  FindVariableFeatures(selection.method = "vst", nfeatures=3000) %>%
  ScaleData() %>%
  SCTransform(vars.to.regress = NULL)

#recluster Immune subset
Immune_harmony <- Immune_harmony  %>%
  RunUMAP(reduction = "rna_harmony", assay = "SCT", dims = 1:50, reduction.name = 'umap.rna', reduction.key = 'rnaUMAP_') %>%
  FindNeighbors(reduction = "rna_harmony") %>%
  FindClusters(resolution = c(0.2, 0.4, 0.6, 0.8, 1.0, 1.2), verbose = TRUE)

##run harmony to remove batch effect again on ATAC assay
DefaultAssay(Immune_harmony) <- "ATAC" 
Immune_harmony <- RunTFIDF(Immune_harmony) 
Immune_harmony <- FindTopFeatures(Immune_harmony, min.cutoff = 'q0') 
Immune_harmony <- RunSVD(Immune_harmony) 
Immune_harmony <- RunUMAP(Immune_harmony, dims = 2:50, reduction = 'lsi')
Immune_harmony <- RunHarmony(object = Immune_harmony, group.by.vars = c('sample_id'),reduction = 'lsi', assay.use = 'ATAC', project.dim = FALSE,  reduction.save = "atac_harmony")
Immune_harmony <- RunUMAP(Immune_harmony, dims = 2:50, reduction = 'atac_harmony', reduction.name = 'umap.atac', reduction.key = 'atacUMAP_')

##to plot Immune_harmony age on RNA DimPlot
DimPlot(Immune_harmony, group.by = "dev.group",reduction = "umap.rna", label = F, cImmunes = devcolor)

##to determine SNN resolution and number of clusters
clustree(Immune_harmony) #determined the SNN resolution @0.6

##to calculate differentially expressed genes (DEG) between different clusters
DefaultAssay(Immune_harmony) <- "RNA"
Immune_harmony_cluster_DEG <- FindAllMarkers(Immune_harmony, only.pos = TRUE, min.pct = 0.25, logfc.threshImmuned = 0.25)

##to calculate differentially accessible peaks (DAP) between different clusters
DefaultAssay(Immune_harmony) <- "ATAC"
Immune_harmony_cluster_DAP <- FindAllMarkers(Immune_harmony, only.pos = TRUE, min.pct = 0.25, logfc.threshImmuned = 0.25)

##based on subcluster DAP, DEG and canonical marker (see Figure 5E analysis), we determined cell type of each cluster
#set active identity to SCT_snn_res.0.6
Immune_harmony <- SetIdent(Immune_harmony, value = Immune_harmony@meta.data$SCT_snn_res.0.6)

#set cluster identity to cell type
{
  cluster_0 <- "Microglia"
  cluster_1 <- "Fetal Microglia"
  cluster_2 <- "Microglia"
  cluster_3 <- "Macrophage"
  cluster_4 <- "Microglia"
  cluster_5 <- "Macrophage"
  cluster_6 <- "Microglia"
  cluster_7 <- "Microglia"
  cluster_8 <- "Immune OL"
  cluster_9 <- "Fetal Microglia"
  cluster_10 <- "Fetal Microglia"
  cluster_11 <- "Lymphatic Endothelial"
  cluster_12 <- "Microglia" #metabolic
  cluster_13 <- "Fetal Microglia"
  cluster_14 <- "Lymphocyte"
  cluster_15 <- "Immune OL"
}

Immune_harmony <- SetIdent(Immune_harmony, value = Immune_harmony@meta.data$SCT_snn_res.0.6)
new.cluster.ids <- c(cluster_0, cluster_1, cluster_2, cluster_3, cluster_4, cluster_5, cluster_6, cluster_7, cluster_8, cluster_9, cluster_10, cluster_11, cluster_12, cluster_13, cluster_14, cluster_15)
names(new.cluster.ids) <- levels(Immune_harmony)
Immune_harmony <- RenameIdents(Immune_harmony, new.cluster.ids)
Immune_harmony$Subcluster <- Immune_harmony@active.ident
Immune_harmony@meta.data$Subcluster <- factor(x = Immune_harmony@meta.data$Subcluster, levels = c("Fetal Microglia", "Microglia", "Macrophage","Lymphocyte","Lymphatic Endothelial", "Immune OL"))

#assign colors to different subclusters
Immune_subcluster_color <- c("Fetal Microglia" = "#FFB3BA", "Microglia" = "#FF6F61","Macrophage" = "#E75480", "Lymphocyte" = "darkred","Lymphatic Endothelial" = "#D5006D", "Immune OL" = "#D8BFD8")

##to plot subcluster on RNA DimPlot
DimPlot(Immune_harmony, label = F, reduction = "umap.rna", cols = Immune_subcluster_color)

##to plot subcluster on ATAC DimPlot
DimPlot(Immune_harmony, label = F, reduction = "umap.atac", cols = Immune_subcluster_color)
```

```{r Figure 5B}
#set active identity to age group
Immune_harmony <- SetIdent(Immune_harmony, value = Immune_harmony@meta.data$dev.group)

##to plot age on RNA DimPlot
DimPlot(Immune_harmony, label = F, reduction = "umap.rna", cols = devcolor)

##to plot age on ATAC DimPlot
DimPlot(Immune_harmony, label = F, reduction = "umap.atac", cols = devcolor)
```

```{r Figure 5C}
##to plot cell type composition difference across age groups on BarPlot
Immune_harmony@meta.data %>% group_by(dev.group, Subcluster) %>% summarise(n = n()) %>% group_by(Subcluster) %>% mutate(freq = n / sum(n)) %>% data.frame -> df1

ggplot(df1, aes(x = Subcluster, y = freq, fill = dev.group)) +
  geom_bar(stat = "identity", width = 0.9) +
  xlab("Subcluster") + ylab("Proportion") + NoLegend() +
  scale_fill_manual(values = dev.group) +
    scale_y_continuous(expand = c(0, 0)) +
    theme_classic() + 
  theme( axis.title.y = element_blank(), axis.title.x = element_blank())
```

```{r Figure 5D}
#subset Microglia
Microglia <- subset(Immune_harmony, idents = c("Microglia", "Fetal Microglia"))

##calculate the DAP in each age group
DefaultAssay(Microglia) <- "ATAC"
Microglia <- SetIdent(Microglia, value = Microglia@meta.data$dev.group)
Microglia <- FindAllMarkers(Microglia, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)

#extract the DAP in each subcluster and make a list
Microglia_age_DAP <- list(Fetal_DAP, Infant_DAP, Child_DAP, Adolescent_DAP, Adult_DAP)

#run the rGreat analysis
Microglia_age_DAP <- lapply(Microglia_age_DAP, function(x) { x <- x %>%
  separate(gene, into = c("seqnames", "start", "end"), sep = "-")
x$start <- as.numeric(x$start)
x$end <- as.numeric(x$end)
x$strand <- c("*")
return(x)})
g.Microglia_age_DAP <- lapply(Microglia_age_DAP, function(x) {x <- GenomicRanges::GRanges(
  seqnames = x$seqnames,
  ranges = IRanges(start = x$start,
                   end = x$end
  ),
  strand = x$strand,
  id = x$cluster,
  rank = x$avg_log2FC,
  adjust = x$p_val_adj)
return(x)})

#calculate the unique DAP in each subcluster
laps_Full <- findOverlapsOfPeaks(g.Microglia_age_DAP[[1]], g.Microglia_age_DAP[[2]], g.Microglia_age_DAP[[3]], g.Microglia_age_DAP[[4]], g.Microglia_age_DAP[[5]])

#visualize the unique DAP in each subcluster using VennDiagram
makeVennDiagram(laps_Full, NameOfPeaks = c("Fetal", "Infant", "Child", "Adolescent", "Adult"), fill=c("Fetal" =  "#0D0887", "Infant" = "#7E03A8", "Child" ="#CC4678","Adolescent" = "#F89441","Adult" = "#F0F921"))

#create a list of age groups for barplot
age <- c("Fetal", "Infant", "Child", "Adolescent", "Adult")

#get the number of unique peaks across age from Immune_harmony_age_DAP
Microglia_age_unique_DAP <- c(12525, 1, 7, 0, 1)

#create a dataframe
Microglia_age_unique_DAP <- data.frame(age, Microglia_age_unique_DAP,stringsAsFactors = FALSE)
Microglia_age_unique_DAP$age <- factor(x = Microglia_age_unique_DAP$age, levels = c("Fetal", "Infant", "Child", "Adolescent", "Adult"))

##to plot unique DAP across age groups on BarPlot
ggplot() +
  geom_bar(data = Microglia_age_unique_DAP, aes(x = age, y = Microglia_age_unique_DAP, fill = age), stat = "identity") + scale_fill_manual(values = devcolor) +
    scale_y_continuous(expand = c(0, 0)) +
    theme_classic() + 
  theme(axis.title.y = element_blank(), axis.title.x = element_blank())
```

```{r Figure 5E}
#set active identity to subcluster
Immune_harmony <- SetIdent(Immune_harmony, value = Immune_harmony@meta.data$Subcluster)
DefaultAssay(Immune_harmony) <- "RNA"

##to plot canonical markers expressed in each subcluster on ViolinPlot
VlnPlot(Immune_harmony, features = c("VIM","APOE","P2RY12","CX3CR1","PTPRC","CD74","HLA-DRA","HLA-DRB1","IL32","LYVE1","PLP1","PTGDS","BCAS1","MAG", "SOX2","OLIG1"), stack = T, flip = T, fill.by = "ident", cols = Immune_subcluster_color)
```

```{r Figure 5F}
#create a subset of only Microglia and Immune OL
subset <- subset(Immune_harmony, idents = c("Fetal Microglia", "Microglia", "Immune Oligo"))

#create a metadata of "Oligo" to indicate whether the cell is either a Microglia or Oligo
subset@meta.data$Oligo <- ifelse(subset@meta.data$Subtype %in% c("Fetal Microglia", "Microglia"), "Microglia", "Oligo")

#set active identity to "Oligo"
DefaultAssay(subset) <- "RNA"
subset <- SetIdent(subset, value = subset@meta.data$Oligo)

#create a metadata of ID_Cell
subset@meta.data$ID_Cell <- gsub(" ", "_", paste0(subset@meta.data$sample_id, "_",
                                subset@meta.data$Oligo))

#get count data
pb <- data.frame(gene = rownames(subset@assays$RNA@counts))

#set rownames as gene names
rownames(pb) <- pb$gene
for (sample in unique(subset@meta.data$ID_Cell)){
    sub <- subset(subset, subset = ID_Cell == sample)
    if (nrow(sub@meta.data) > 9){
        pb[sample] <- rowSums(sub@assays$RNA@counts)
    } else {
        print(paste("NOTE: sample", sample, "has less than 10 cells. 
        This sample will be skipped."))
    }
}

#remove gene column (since rownames are genes)
pb <- pb[, -1]
colData <- data.frame(sample = colnames(pb))

#create a column that is just the cell type
colData$cellType <- ifelse(grepl("Microglia", colData$sample), "Microglia", "Oligo")

#set contrast
contrast <- c("cellType", "Oligo", "Microglia")

# set rownames as the sample names
rownames(colData) <- colData$sample

#run DESeq2
dds <- DESeq2::DESeqDataSetFromMatrix(
            countData = pb,
            colData = colData,
            design = ~ cellType)
dds <- DESeq2::DESeq(dds)
rld <- rlog(dds, blind=TRUE)
DESeq2::plotPCA(rld, intgroup = "cellType")

#resultsNames(dds)
res <- results(dds, 
               contrast = contrast,
               alpha = 0.05)
res <- lfcShrink(dds,
                 contrast =  contrast,
                 res=res, type = "normal")
res_tbl <- res %>%
        data.frame() %>%
        rownames_to_column(var="gene") %>%
        as_tibble()

# Set thresholds
padj_cutoff <- 0.05
res_table_thres <- res_tbl %>% 
                  mutate(threshold = padj < 0.05 & abs(log2FoldChange) >= 1)

#set comparison direction
res_table_thres$direction <- ""
res_table_thres$direction[!is.na(sign(res_table_thres$log2FoldChange))] <- ifelse(sign(res_table_thres$log2FoldChange[!is.na(sign(res_table_thres$log2FoldChange))])==-1, yes = "down", no = "up")

#sum(res_table_thres$threshold & (res_table_thres$direction=="up"))
table(res_table_thres$threshold, res_table_thres$direction)

##to plot the VolcanoPlot of DEG between Immune Oligo and Microglia
EnhancedVolcano(res_table_thres,
    lab = res_table_thres$gene,
    x = 'log2FoldChange',
    y = 'padj',
    pCutoff = 0.05,  ylim = c(0,5),
    selectLab = c("MEF2","AKAP13","BCAS1","MBP","UGT8"),
    drawConnectors = TRUE ,legendPosition = 'right',title = NULL,
    subtitle = NULL,
    col=c('black', 'black', 'black', 'red3')) + NoLegend()
```

```{r Figure 5G}
#set sobj identity to subcluster (we already organized the subcluster info of RBFOX1 OL, OPALIN OL , Immune OL into the sobj)
sobj <- SetIdent(sobj, value = sobj@meta.data$Subcluster)

#create a subset of only OL and Immune OL
subset <- subset(sobj, idents = c("RBFOX1 OL", "OPALIN OL" , "Immune OL"))

#create a metadata of "Oligo" to indicate whether the cell is either a OL or Immune OL
subset@meta.data$Oligo <- ifelse(subset@meta.data$Subtype %in% c("RBFOX1 Oligo", "OPALIN Oligo"), "Oligo", "Immune Oligo")

#set active identity to "Oligo"
DefaultAssay(subset) <- "RNA"
subset <- SetIdent(subset, value = subset@meta.data$Oligo)

#create a metadata of ID_Cell
subset@meta.data$ID_Cell <- gsub(" ", "_", paste0(subset@meta.data$sample_id, "_",
                                subset@meta.data$Oligo))

#get count data
pb <- data.frame(gene = rownames(subset@assays$RNA@counts))

#set rownames as gene names
rownames(pb) <- pb$gene
for (sample in unique(subset@meta.data$ID_Cell)){
    sub <- subset(subset, subset = ID_Cell == sample)
    if (nrow(sub@meta.data) > 9){
        pb[sample] <- rowSums(sub@assays$RNA@counts)
    } else {
        print(paste("NOTE: sample", sample, "has less than 10 cells. 
        This sample will be skipped."))
    }
}

#remove gene column (since rownames are genes)
pb <- pb[, -1]
colData <- data.frame(sample = colnames(pb))

#create a column that is just the cell type
colData$cellType <- ifelse(grepl("Immune_Oligo", colData$sample), "Immune Oligo", "Oligo")

#set contrast
contrast <- c("cellType", "Immune Oligo", "Oligo")

# set rownames as the sample names
rownames(colData) <- colData$sample

#run DESeq2
dds <- DESeq2::DESeqDataSetFromMatrix(
            countData = pb,
            colData = colData,
            design = ~ cellType)
dds <- DESeq2::DESeq(dds)
rld <- rlog(dds, blind=TRUE)
DESeq2::plotPCA(rld, intgroup = "cellType")

#resultsNames(dds)
res <- results(dds, 
               contrast = contrast,
               alpha = 0.05)
res <- lfcShrink(dds,
                 contrast =  contrast,
                 res=res, type = "normal")
res_tbl <- res %>%
        data.frame() %>%
        rownames_to_column(var="gene") %>%
        as_tibble()

# Set thresholds
padj_cutoff <- 0.05
res_table_thres <- res_tbl %>% 
                  mutate(threshold = padj < 0.05 & abs(log2FoldChange) >= 1)

#set comparison direction
res_table_thres$direction <- ""
res_table_thres$direction[!is.na(sign(res_table_thres$log2FoldChange))] <- ifelse(sign(res_table_thres$log2FoldChange[!is.na(sign(res_table_thres$log2FoldChange))])==-1, yes = "down", no = "up")

#sum(res_table_thres$threshold & (res_table_thres$direction=="up"))
table(res_table_thres$threshold, res_table_thres$direction)

##to plot the VolcanoPlot of DEG between Immune OL and OL
EnhancedVolcano(res_table_thres,
    lab = res_table_thres$gene,
    x = 'log2FoldChange',
    y = 'padj',
    pCutoff = 0.05,  ylim = c(0,5),
    selectLab = c("ATP8B4","PALD1","CYBB"),
    drawConnectors = TRUE ,legendPosition = 'right',title = NULL,
    subtitle = NULL,
    col=c('black', 'black', 'black', 'red3')) + NoLegend()
```

```{r Figure 5H}
##calculate DEG between different cell types
DefaultAssay(OPC_OL) <- "RNA"
OPC_OL_cell_DEG <- FindAllMarkers(OPC_OL, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)

#extract DEG of each cell type from OPC_OL_cell_DEG and make a list:
list <- c("Fetal COP","OPC", "COP", "Early OL", "OL")

##Load the genesets
ensembl = useMart("ensembl", dataset = "hsapiens_gene_ensembl")

# For C5 gene ontology genesets
C5genesets <- msigdbr(species = "Homo sapiens", category = "C5")
C5togene <- C5genesets[,c("gs_exact_source","gene_symbol")]
C5toname <- C5genesets[,c("gs_exact_source","gs_name")]

# For GO Biological Process
GOgenesets <- subset(C5genesets, subset=C5genesets[,"gs_subcat"]=="GO:BP")
GOtogene <- GOgenesets[,c("gs_exact_source","gene_symbol")]
GOtoname <- GOgenesets[,c("gs_exact_source","gs_name")]

##calculate DEG across age groups in Microglia subset
DefaultAssay(Microglia) <- "RNA"
Microglia <- SetIdent(Microglia, value = Microglia@meta.data$dev.group)
Microglia_age_DEG <- FindAllMarkers(Microglia, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)

#extract DEG of each age group from Microglia_age_DEG and make a list:
list <- c("Fetal", "Infant", "Child", "Adolescent", "Adult")

#get the GO Biological Process
GO <- lapply(list, function (x)
{
  print(x)
  y <- enricher(gene = rownames(eval(parse(text = x))),
                pvalueCutoff = 0.05,
                pAdjustMethod = "BH",
                minGSSize = 15,
                maxGSSize = 1000,
                qvalueCutoff = 0.2,
                TERM2GENE = GOtogene,
                TERM2NAME = GOtoname)
}
)

#GO Top5 Heatmap
names(GO) <- list
Desc_list <- lapply(list, function(x){
df <- GO[[x]]
df <- df[df$p.adjust < 0.05,]
df <- df[order(df$p.adjust, decreasing = FALSE),]
desc <- df$Description
desc <- desc[1:5]
return(desc)
}
)
Desc <- unlist(Desc_list) %>% unique()
Desc.df <- as.data.frame(Desc)
Desc.df <- na.omit(Desc.df)
rownames(Desc.df) <- Desc.df$Desc
Desc <- na.omit(Desc)
cols <-lapply(list, function(age){
  Desc.df[age] <- 0
  Cell.Type.Desc.df <- data.frame(GO[[age]])
  Cell.Type.Desc.df$qvalue.scaled <- -log(Cell.Type.Desc.df$qvalue)
  sig <-data.frame(Cell.Type.Desc.df[Cell.Type.Desc.df$p.adjust < 0.05,]) 
  sig <- sig[order(sig$p.adjust, decreasing = F),]
  Desc.df[sig$Description,age]<- Cell.Type.Desc.df[Cell.Type.Desc.df$Description %in% sig$Description, "qvalue.scaled"]
  Desc.df <- Desc.df[unique(Desc),] 
  rownames(Desc.df) <- Desc.df$Desc
  Desc.df$Desc <- NULL
  return(Desc.df)
  })
df2 <- data.frame(cols)
matrix_GO <- as.matrix(df2)
colnames(matrix_GO) <- colnames(df2)
rownames(matrix_GO) <- unique(Desc)
matrix_GO = t(apply(df2, 1, scale))

# Annotation
AnnoDF <- data.frame(c("Fetal", "Infant", "Child", "Adolescent", "Adult"))
colnames(AnnoDF) <- c("Age")
AnnoDF$Age <- factor(AnnoDF$Age, levels = c("Fetal", "Infant", "Child", "Adolescent", "Adult"))
colAnno <- HeatmapAnnotation(
  df=AnnoDF,
  show_annotation_name = c(TRUE,TRUE),
  annotation_name_gp = gpar(fontsize = 12),
  annotation_name_side = "left",
  col = list(Age = c("Fetal" =  "#0D0887", "Infant" = "#7E03A8", "Child" ="#CC4678","Adolescent" = "#F89441","Adult" = "#F0F921")))

#Draw GO heatmap
plot_GO <- Heatmap(matrix_GO, cluster_columns=F,cluster_rows =FALSE, row_order = Desc, row_names_gp = gpar(fontsize = 5),top_annotation = colAnno, show_heatmap_legend = F)
draw(plot_GO, heatmap_legend_side="left", annotation_legend_side="left")
```











