---
title: "Supplemental_Figuer_5_Analysis"
author: "Sheldon Ding"
date: "2025-03-27"
output: html_document
---

```{r load libraries}
#packages (some packages might require dependent packages)
library(Matrix)
library(Seurat)
library(Signac)
library(patchwork)
library(tidyverse)
library(dplyr)
library(readr)
library(magrittr)
library(openxlsx) 
library(harmony)
library(clustree)
library(rGREAT)
library(GenomicRanges)
library(msigdbr)
library(clusterProfiler)
library(AnnotationHub)
library(TFBSTools)
library(BSgenome.Hsapiens.UCSC.hg38)
library(IRanges)
library(Biostrings)
library(openxlsx)
library(biomaRt)
library(DESeq2)
library(grid)
library(SeuratWrappers)
library(monocle3)
library(EnhancedVolcano)
```

```{r set seed}
set.seed(6)
```

```{r load object}
OL_harmony <- readRDS("output/objects/OL_harmony.RDS")
```

```{r set color for cell type and age}
#cell color
Immune_subcluster_color <- c("Fetal Microglia" = "#FFB3BA", "Microglia" = "#FF6F61","Macrophage" = "#E75480", "Lymphocyte" = "darkred","Lymphatic Endothelial" = "#D5006D", "Immune OL" = "#D8BFD8")

#age
devcolor <- c("Fetal" =  "#0D0887", "Infant" = "#7E03A8", "Child" ="#CC4678","Adolescent" = "#F89441","Adult" = "#F0F921")
```


###Supplemental Figure 5
```{r Supplemental Figure 5A}
#set active identity to Subcluster
Immune_harmony <- SetIdent(Immune_harmony, value = Immune_harmony@meta.data$Subcluster)

##to plot Subcluster on RNA DimPlot
DimPlot(Immune_harmony, label = T, reduction = "umap.rna", cols = Immune_subcluster_color)

##to plot expression of transcripts associated with immune markers, transcription factors identified in oligodendrocyte lineage cells, and myelin in the immune oligodendrocyte population on FeaturePlot
DefaultAssay(Immune_harmony) <- "RNA"
FeaturePlot(Immune_harmony, features = c("CD74","CX3CR1","SOX2", "MYRF", "PLP1","MOBP"), keep.scale = "all")
```

```{r Supplemental Figure 5B}
##compute the GC content for each peak
DefaultAssay(Immune_harmony) <- "ATAC"
Immune_harmony <- RegionStats(Immune_harmony, genome = BSgenome.Hsapiens.UCSC.hg38)

##link peaks to genes
#x = "CD74","CX3CR1","SOX2","OLIG1","PLP1","MOBP" (use one gene at a time to link peaks to genes and do the CoveragePlot)
Immune_harmony <- LinkPeaks(
  object = Immune_harmony,
  peak.assay = "ATAC",
  expression.assay = "RNA",
  genes.use = x)

##to visualize open peaks, RNA expression and links on CoveragePlot
CoveragePlot(
  object = Immune_harmony,
  region = x,
  features = x,
  expression.assay = "RNA",
  ) & scale_fill_manual(values = Immune_subcluster_color)
```








