---
title: "Supplemental_Figuer_4_Analysis"
author: "Sheldon Ding"
date: "2025-03-27"
output: html_document
---

```{r load libraries}
#packages (some packages might require dependent packages)
library(Matrix)
library(Seurat)
library(Signac)
library(patchwork)
library(tidyverse)
library(dplyr)
library(readr)
library(magrittr)
library(openxlsx) 
library(harmony)
library(clustree)
library(rGREAT)
library(GenomicRanges)
library(msigdbr)
library(clusterProfiler)
library(AnnotationHub)
library(TFBSTools)
library(BSgenome.Hsapiens.UCSC.hg38)
library(IRanges)
library(Biostrings)
library(openxlsx)
library(biomaRt)
library(DESeq2)
library(grid)
library(SeuratWrappers)
library(monocle3)
library(EnhancedVolcano)
```

```{r set seed}
set.seed(6)
```

```{r load object}
OL_harmony <- readRDS("output/objects/OL_harmony.RDS")
```

```{r set color for cell type and age}
#age
devcolor <- c("Fetal" =  "#0D0887", "Infant" = "#7E03A8", "Child" ="#CC4678","Adolescent" = "#F89441","Adult" = "#F0F921")
```


###Supplemental Figure 4
```{r Supplemental Figure 4A}
#set active identity to SCT_snn_res.0.6
OL_harmony <- SetIdent(OL_harmony, value = OL_harmony@meta.data$SCT_snn_res.0.6)

##to plot cluster on RNA DimPlot
DimPlot(OL_harmony, label = T, reduction = "umap.rna")
```

```{r Supplemental Figure 4B}
##to plot differentially expressed canonical genes on FeaturePlot
DefaultAssay(OL_harmony) <- "RNA" 
FeaturePlot(OL_harmony, features = c("PDGFRA","PLP1", "MBP","C10orf90","CSPG4","OPALIN","RBFOX1","PTGDS"), reduction = "umap.rna",keep.scale = "all")
```

```{r Supplemental Figure 4C}
##calculate the DAP in each subcluster
DefaultAssay(OL_harmony) <- "ATAC"
OL_harmony <- SetIdent(OL_harmony, value = OL_harmony@meta.data$Subcluster)
OL_harmony_subcluster_DAP <- FindAllMarkers(OL_harmony, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)

#extract the DAP in each subcluster and make a list
OL_harmony_subcluster_DAP_list <- list(Early_OL_DAP,RBFOX1_OL_DAP,OPALIN_OL_DAP)

#run the rGreat analysis
OL_harmony_subcluster_DAP_list <- lapply(OL_harmony_subcluster_DAP_list, function(x) { x <- x %>%
  separate(gene, into = c("seqnames", "start", "end"), sep = "-")
x$start <- as.numeric(x$start)
x$end <- as.numeric(x$end)
x$strand <- c("*")
return(x)})
g.OL_harmony_subcluster_DAP_list <- lapply(OL_harmony_subcluster_DAP_list, function(x) {x <- GenomicRanges::GRanges(
  seqnames = x$seqnames,
  ranges = IRanges(start = x$start,
                   end = x$end
  ),
  strand = x$strand,
  id = x$cluster,
  rank = x$avg_log2FC,
  adjust = x$p_val_adj)
return(x)})

#rGreat analysis for computing the differenet pathway based on DAP
r.OL_harmony_subcluster_DAP_list <- rGREAT::great(g.OL_harmony_subcluster_DAP_list,
                        gene_sets = "msigdb:C5:GO:BP",
                        tss_source = "hg38"
                        )
ETable.subcluster.list <- rGREAT::getEnrichmentTable(r.OL_harmony_subcluster_DAP_list)

#extract GO pathway for RBFOX1
RGA.RBFOX1 <- ETable.subcluster.list[[2]]
RGA.RBFOX1_gene_names <- unstrsplit(RGA.RBFOX1$annotated_genes, sep = " ")
RGA.RBFOX1_gene_names <- scan(text = RGA.RBFOX1_gene_names, what = "")

RBFOX1 <- enricher(gene = RGA.RBFOX1_gene_names,
              pvalueCutoff = 0.05,
              pAdjustMethod = "BH",
              minGSSize = 15,
              maxGSSize = 1000,
              qvalueCutoff = 0.2,
              TERM2GENE = GOtogene,
              TERM2NAME = GOtoname)

##to plot top expressed biological processes for RBFOX1+ oligodendrocyte clusters on DotPlot
dotplot(RBFOX1, showCategory=20)
```

```{r Supplemental Figure 4D}
RGA.OPALIN <- ETable.subcluster.list[[3]]
RGA.OPALIN_gene_names <- unstrsplit(RGA.OPALIN$annotated_genes, sep = " ")
RGA.OPALIN_gene_names <- scan(text = RGA.OPALIN_gene_names, what = "")

OPALIN <- enricher(gene = RGA.OPALIN_gene_names,
              pvalueCutoff = 0.05,
              pAdjustMethod = "BH",
              minGSSize = 15,
              maxGSSize = 1000,
              qvalueCutoff = 0.2,
              TERM2GENE = GOtogene,
              TERM2NAME = GOtoname)

##to plot top expressed biological processes for OPALIN+ oligodendrocyte clusters on DotPlot
dotplot(OPALIN, showCategory=20)
```

```{r Supplemental Figure 4E}
##calculate the DAP in each age group
DefaultAssay(OL_harmony) <- "ATAC"
OL_harmony <- SetIdent(OL_harmony, value = OL_harmony@meta.data$dev.group)
OL_harmony_age_DAP <- FindAllMarkers(OL_harmony, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)

#extract the DAP in each subcluster and make a list
OL_harmony_age_DAP <- list(Fetal_DAP, Infant_DAP, Child_DAP, Adolescent_DAP, Adult_DAP)

#run the rGreat analysis
OL_harmony_age_DAP <- lapply(OL_harmony_age_DAP, function(x) { x <- x %>%
  separate(gene, into = c("seqnames", "start", "end"), sep = "-")
x$start <- as.numeric(x$start)
x$end <- as.numeric(x$end)
x$strand <- c("*")
return(x)})
g.OL_harmony_age_DAP_list <- lapply(OL_harmony_age_DAP, function(x) {x <- GenomicRanges::GRanges(
  seqnames = x$seqnames,
  ranges = IRanges(start = x$start,
                   end = x$end
  ),
  strand = x$strand,
  id = x$cluster,
  rank = x$avg_log2FC,
  adjust = x$p_val_adj)
return(x)})

#rGreat analysis for computing the differenet pathway based on DAP
r.OL_harmony_age_DAP_list <- rGREAT::great(g.OL_harmony_age_DAP_list,
                        gene_sets = "msigdb:C5:GO:BP",
                        tss_source = "hg38"
                        )
ETable.age.list <- rGREAT::getEnrichmentTable(r.OL_harmony_age_DAP_list)

#make annotation dataframe and DF for heatmap
AnnoDF <- data.frame(c("Fetal","Infant", "Child","Adolescent", "Adult"))
colnames(AnnoDF) <- c("Age")
AnnoDF$Age <- factor(AnnoDF$Age, levels = c("Fetal","Infant", "Child","Adolescent", "Adult"))
colAnno <- HeatmapAnnotation(
  df=AnnoDF,
  show_annotation_name = c(TRUE,TRUE),
  annotation_name_gp = gpar(fontsize = 12),
  annotation_name_side = "left",
  col = list(Age = c("Fetal" =  "#0D0887", "Infant" = "#7E03A8", "Child" ="#CC4678","Adolescent" = "#F89441","Adult" = "#F0F921")))

names(GO) <- list
Desc_list <- lapply(list, function(x){
df <- GO[[x]]
df <- df[df$p_value < 0.05,]
df <- df[order(df$p_value, decreasing = FALSE),]
desc <- df$id
desc <- desc[1:5]
return(desc)
}
)

Desc <- unlist(Desc_list) %>% unique()
Desc.df <- as.data.frame(Desc)
Desc.df <- na.omit(Desc.df)
rownames(Desc.df) <- Desc.df$Desc
Desc <- na.omit(Desc)
cols <-lapply(list, function(great){
  Desc.df[great] <- 0
  great.Desc.df <- data.frame(GO[[great]])+
  sig <-data.frame(great.Desc.df[great.Desc.df$p_value < 0.05,]) 
  sig <- sig[order(sig$p_value, decreasing = F),]
  Desc.df[sig$id,great]<- great.Desc.df[great.Desc.df$id %in% sig$id, "fold_enrichment"]
  Desc.df <- Desc.df[unique(Desc),] 
  rownames(Desc.df) <- Desc.df$Desc
  Desc.df$Desc <- NULL
  return(Desc.df)
  })
df2 <- data.frame(cols)
matrix_GO <- as.matrix(df2)
colnames(matrix_GO) <- colnames(df2)
rownames(matrix_GO) <- unique(Desc)
matrix_GO = t(apply(df2, 1, scale))

##to plot top gene ontology biological processes identified by GREAT analysis of differentially accessible peaks in oligodendrocytes from each age group on Heatmap
plot_GO <- Heatmap(matrix_GO, cluster_columns=F,cluster_rows =FALSE
                       , row_order = Desc, row_names_gp = gpar(fontsize = 5),top_annotation = colAnno, show_column_names = TRUE)
draw(plot_GO)
```

```{r Supplemental Figure 4E}

```















